### ISSUE TITLE
LinqToDB.SqlQuery.SqlException: 'Table not found for 'LinqToDB.SqlQuery.SqlColumn'.' when using Contains

### ISSUE THREAD

**Issue by vladopandzic** [linq2db/linq2db#2608]
Problem is when I try to use Contains method when comparing each member of list to some filed in database

For example:
Part of query only where exception is thrown:

```
  var a = (from q in query where !faze.Contains(q.StatusPhase) select new { q.UnitId, q.FromDate, q.ToDate }).GroupBy(x => new { x.UnitId, x.FromDate, x.ToDate }).Select(x => new
                    {
                        // count = x.Count()
                    }).Count();
```

where `faze `is variable as defined:

            var faze = new List<int> { 11, 18, 19, 20, 21, 22, 23, 24, 26, 29, 28 };
          

```
Exception message:
LinqToDB.SqlQuery.SqlException: 'Table not found for 'LinqToDB.SqlQuery.SqlColumn'
Stack trace:
```
```
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildExpression(ISqlExpression expr, Boolean buildTableName, Boolean checkParentheses, String alias, Boolean& addAlias, Boolean throwExceptionIfTableNotFound)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildExpression(Int32 parentPrecedence, ISqlExpression expr, String alias, Boolean& addAlias)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildInListValues(InList predicate, IEnumerable values)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildInListPredicate(ISqlPredicate predicate)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildPredicate(ISqlPredicate predicate)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildSearchCondition(SqlSearchCondition condition)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildSearchCondition(Int32 parentPrecedence, SqlSearchCondition condition)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildPredicate(ISqlPredicate predicate)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildSearchCondition(SqlSearchCondition condition)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildSearchCondition(Int32 parentPrecedence, SqlSearchCondition condition)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildWhereClause(SelectQuery selectQuery)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildSelectQuery(SqlSelectStatement selectStatement)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildSql(Int32 commandNumber, SqlStatement statement, StringBuilder sb, Int32 indent, Boolean skipAlias)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildSqlBuilder(SelectQuery selectQuery, Int32 indent, Boolean skipAlias)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildPhysicalTable(ISqlTableSource table, String alias)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildTableName(SqlTableSource ts, Boolean buildName, Boolean buildAlias)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildFromClause(SqlStatement statement, SelectQuery selectQuery)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildSelectQuery(SqlSelectStatement selectStatement)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildSql(Int32 commandNumber, SqlStatement statement, StringBuilder sb, Int32 indent, Boolean skipAlias)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildSql(Int32 commandNumber, SqlStatement statement, StringBuilder sb, Int32 startIndent)
   at LinqToDB.Data.DataConnection.QueryRunner.GetCommand(DataConnection dataConnection, IQueryContext query, Int32 startIndent)
   at LinqToDB.Data.DataConnection.QueryRunner.SetQuery()
   at LinqToDB.Linq.QueryRunnerBase.SetCommand(Boolean clearQueryHints)
   at LinqToDB.Data.DataConnection.QueryRunner.ExecuteReader()
   at LinqToDB.Linq.QueryRunner.ExecuteElement[T](Query query, IDataContext dataContext, Mapper`1 mapper, Expression expression, Object[] ps, Object[] preambles)
   at LinqToDB.Linq.QueryRunner.<>c__DisplayClass26_0`1.<SetRunQuery>b__0(IDataContext db, Expression expr, Object[] ps, Object[] preambles)
   at LinqToDB.Linq.ExpressionQuery`1.System.Linq.IQueryProvider.Execute[TResult](Expression expression)
   at WebRenter.Repository.ReservationsRepository.FilterReservations(ReservationCriteria criteria, IQueryable`1& query, Int32& count) in C:\PROJEKTI\Adriagate\WebSites\WebRenter\WebRenter.Repository\ReservationsRepository.cs:line 826
   at WebRenter.Repository.ReservationsRepository.GetReservations(PrimaryKey userPk, AdriagateDatabase db, Int32& count, ReservationCriteria criteria) in C:\PROJEKTI\Adriagate\WebSites\WebRenter\WebRenter.Repository\ReservationsRepository.cs:line 676
   at WebRenter.Repository.ReservationsRepository.GetAllReservations(PrimaryKey userPk, ReservationCriteria criteria) in C:\PROJEKTI\Adriagate\WebSites\WebRenter\WebRenter.Repository\ReservationsRepository.cs:line 426
   at WebRenter.Controllers.API.ReservationsController.Reservations(ReservationCriteria criteria) in C:\PROJEKTI\Adriagate\WebSites\WebRenter\WebRenter.Services\API\ReservationsController.cs:line 115
   at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ActionExecutor.<>c__DisplayClass13.<GetExecutor>b__c(Object instance, Object[] methodParameters)
   at System.Threading.Tasks.TaskHelpers.RunSynchronously[TResult](Func`1 func, CancellationToken cancellationToken)

```
### Steps to reproduce
Somewhat complex query that contains "Contains" method in it

This happened after we converted out BL Toolkit to linq to db

EDIT:
When I try to write it like this I get exception that it can not be converted to SQL.

```
   var a = (from q in query
                             where !(faze.Any(f => f == q.StatusPhase))
                             select new { q.UnitId, q.FromDate, q.ToDate }).GroupBy(x => new { x.UnitId, x.FromDate, x.ToDate }).Select(x => new
                    {
                        // count = x.Count()
                    }).Count();

```
### Environment details
linq2db version: *3.1.3*
Database Server: *Sql Server *
Database Provider: *SqlServer.2012*
Operating system: *Windows 10*
.NET Framework: *4.6.1*

**Comment by sdanyliv** [comment]
Try these queries:
```cs
var a = 
   from q in query 
   where !faze.Contains(q.StatusPhase) 
   group q by new { q.UnitId, q.FromDate, q.ToDate } into g
   select new 
   {
        count = g.Count()
   };
```
```cs
var a = 
   (from q in query 
   where !faze.Contains(q.StatusPhase) 
   select new { q.UnitId, q.FromDate, q.ToDate })
   .Distinct()
   .Count()
```
Last query that you have posted is not convertible to the SQL and you have proper exception.

**Comment by vladop** [comment]
Nope, I tryed that queries now. And other with Any method or All method.
However this seems to works:

  ```
var a = (from q in query
                             let phases = new int[] { 11, 18, 19, 20, 21, 22, 23, 24, 26, 29, 28 }
                             where phases.All(f => f != q.StatusPhase)
                             select new { q.UnitId, q.FromDate, q.ToDate }).GroupBy(x => new { x.UnitId, x.FromDate, x.ToDate }).Select(x => new
                             {

                             }).Count();
```
However it does not work if variable `phases  `is defined elsewhere (I need it as variable, to avoid duplication)
This also not work` ... let phases=faze ... `(where faze is defined elsewhere)

**Comment by sdanyliv** [comment]
Will check that, looks like a bug. Your last query woks because of different functionality used in query.
Non working query is very simple. Probably there is something in `query` which you have not shown in this issue.

**Comment by sdanyliv** [comment]
@vladop, please also check that issue is reproducible in latest version.

**Comment by vladop** [comment]
yes, it is reproducable using 3.1.6

**Comment by sdanyliv** [comment]
@vladop, could you please post full query?

**Comment by vladop** [comment]
It is long query but this some first part of it without filtering logic:

```
  var query = (from ugovori in db.SMJUgovori.Where(x => x.Deleted == false && ((x.DobavljaciCode == userPk.Code && x.DobavljaciSite == userPk.Site) || userPk.Code == 0))
                         join o in db.SMJObjekti.Where(x => x.Deleted == false) on new { ugovori.ObjektiCode, ugovori.ObjektiSite } equals new { o.ObjektiCode, o.ObjektiSite } into oo
                         from o in oo
                         from ug in
                             (from u2 in db.SMJUgovori.Where(x => x.Deleted == false && o.ObjektiCode == x.ObjektiCode && ((x.DobavljaciCode == userPk.Code && x.DobavljaciSite == userPk.Site) || userPk.Code == 0)) select new { u2.DatumOd, u2.UgovoriCode, u2.UgovoriSite, u2.DobavljaciCode, u2.DobavljaciSite }).OrderByDescending(x => x.DatumOd).Take(userPk.Code == 0 ? 20 : 1)
                             //   (from o in db.SMJObjekti.Where(x => x.Deleted == false)
                         join PREF in db.SMJPrefiksi.Where(x => x.Deleted == false) on new { c = o.PrefiksiCode, s = o.PrefiksiSite } equals new { c = (int?)PREF.PrefiksiCode, s = (int?)PREF.PrefiksiSite } into pref1
                         from PREF in pref1.DefaultIfEmpty()
                             //from ug in
                             //      (from u2 in db.SMJUgovori.Where(x => x.Deleted == false && o.ObjektiCode == x.ObjektiCode && ((x.DobavljaciCode == userPk.Code && x.DobavljaciSite == userPk.Site) || userPk.Code == 0)) select new { u2.DatumOd, u2.UgovoriCode, u2.UgovoriSite, u2.DobavljaciCode, u2.DobavljaciSite }).OrderByDescending(x => x.DatumOd).Take(userPk.Code == 0 ? 1000 : 1)

                         join p in db.OSPPartneri.Where(x => x.Deleted == false) on new { c = ug.DobavljaciCode, s = ug.DobavljaciSite } equals new { c = p.PartneriCode, s = p.PartneriSite }
                         join u in db.SMJSmjestajneJedinice.Where(x => x.Deleted == false) on new { o.ObjektiCode, o.ObjektiSite } equals new { u.ObjektiCode, u.ObjektiSite }
                         join ugosta in db.SMJUgovoriStavke on new { u.SmjestajneJediniceCode, u.SmjestajneJediniceSite, ug.UgovoriCode, ug.UgovoriSite, Deleted = false } equals new { ugosta.SmjestajneJediniceCode, ugosta.SmjestajneJediniceSite, ugosta.UgovoriCode, ugosta.UgovoriSite, ugosta.Deleted }
                         join r in db.SMJKalkulacije on new { c = u.SmjestajneJediniceCode, s = u.SmjestajneJediniceSite, Deleted = false } equals new { c = r.SmjestajneJediniceCode.Value, s = r.SmjestajneJediniceSite.Value, r.Deleted }
                         join rp in db.BKGRezervacije on new { r.RezervacijeCode, r.RezervacijeSite, r.Deleted } equals new { rp.RezervacijeCode, rp.RezervacijeSite, rp.Deleted }
                         join g in db.OSPPartneri on new { c = rp.GostiCode.Value, s = rp.GostiSite.Value, Deleted = false } equals new { c = g.PartneriCode, s = g.PartneriSite, g.Deleted }
                         join f in db.OSPVrsteStatusaKalkulacije on new { r.VrsteStatusaKalkulacijeCode, r.VrsteStatusaKalkulacijeSite } equals new { f.VrsteStatusaKalkulacijeCode, f.VrsteStatusaKalkulacijeSite }
                         join fj in db.OSPVrsteStatusaKalkulacijeJezici.Where(x => (x.Tip == 2/* || x.Tip==0*/) && x.JeziciCode == 1 && x.JeziciSite == 61) on new { f.VrsteStatusaKalkulacijeCode, f.VrsteStatusaKalkulacijeSite } equals new { fj.VrsteStatusaKalkulacijeCode, fj.VrsteStatusaKalkulacijeSite }

                         join pk in db.OSPPartneriKartice.Where(x => x.Deleted == false && x.TipVeze == "REZ") on new { a = (int?)r.RezervacijeCode, b = (int?)r.RezervacijeSite } equals new { a = pk.VezaCode, b = pk.VezaSite } into pk1
                         from smjkar in pk1.DefaultIfEmpty()
                         join pp in db.OSPPartneri.Where(x => x.Deleted == false && x.RootPartneriId == homeAwayPartnerRootId) on new { c = r.AffiliateCode, d = r.AffiliateSite } equals new { c = (int?)pp.PartneriCode, d = (int?)pp.PartneriSite } into pp1
                         from pp in pp1.DefaultIfEmpty()
                         where ((p.PartneriCode == userPk.Code && p.PartneriSite == userPk.Site) || userPk.Code == 0)

                         select new ReservationModel
                         {
                             AgencyName = !string.IsNullOrEmpty(pp.Ime) && short.Parse(f.Oznaka) == 70 && p != null && p.RootPartneriId == homeAwayPartnerRootId ? pp.Ime : null,
                             Code = r.KalkulacijeCode,
                             Site = r.KalkulacijeSite,
                             RealReservationID = rp.RezervacijeCode + "-" + rp.RezervacijeSite,
                             ReservationID = r.KalkulacijeCode + "-" + r.KalkulacijeSite,
                             CalculationNumber = r.BrojKalkulacijeFull,
                             StatusPhase = short.Parse(f.Oznaka),
                             StatusDescription = fj.Naziv,
                             StatusDetails = f.Opis,
                             ReservationNumber = u.Sifra,
                             GrownUps = r.BrojOdraslih,
                             Teens = r.BrojTeens,
                             Children = r.BrojDjece,
                             Babies = r.BrojBeba,
                             ReservationDate = r.Datum.Value,
                             RenterOptionDate = r.DatumOpcijeIzn,
                             RenterOptionDateSecond = r.DatumOpcijeIzn,
                             ToDate = r.DatumDo,
                             FromDate = r.DatumOd,
                             Price = 0,

                             GuestOptionDate = f.Oznaka == "30" || f.Oznaka == "40" || f.Oznaka == "41" || f.Oznaka == "42" ? r.DatumOpcijeIzn : null,
                             GuestCountry = g.Drzava.Naziv,
                             GuestCountryOznaka = g.Drzava.Sifra,
                             GuestName = (g.Ime + " " + g.Prezime).Trim(),

                             GuestTelephone = (r.DatumOd.Date.AddDays(-7) <= today && today <= r.DatumOd.Date.AddDays(2)) ? g.Telefon : null,
                             GuestMobile = (r.DatumOd.Date.AddDays(-7) <= today && today <= r.DatumOd.Date.AddDays(2)) ? (g.Mobitel1 != null ? g.Mobitel1 : g.Mobitel2) : null,
                             GuestExpectedToArriveHour = r.VrijemeDolaska,
                             GuestPkCode = g.PartneriCode,
                             GuestPkSite = g.PartneriSite,
                             DateInserted = r.Datum,
                             GuestCountryCode = g.DrzaveCode,
                             GuestCountrySite = g.DrzaveSite,

                             ObjectName = o.Naziv,
                             ObjectPrefix = PREF.Naziv,
                             ObjectId = o.ObjektiCode + "-" + o.ObjektiSite,
                             ObjectDC = o.Sifra,
                             ObjectRenterDC = u.SifraIzn,

                             UnitDC = u.Sifra,
                             UnitId = u.SmjestajneJediniceCode + "-" + u.SmjestajneJediniceSite,
                             ObjectCode = o.ObjektiCode,//json ignored.....
                             ObjectSite = o.ObjektiSite,
                             UnitCode = u.SmjestajneJediniceCode,
                             UnitSite = u.SmjestajneJediniceSite,
                             UnitRenterDC = u.SifraIzn,
                             UnitMaxPersons = u.MaxOsoba,
                             UnitClassification = u.Oznaka,
                             Note = r.Napomena,
                             RenterNote = "",
                             HasLinkedCreditCard = (smjkar.VezaCode != null && short.Parse(f.Oznaka) == 70),
                             CreditCardCode = smjkar.VezaCode,
                             CreditCardSite = smjkar.VezaSite,
                             IsCalculation = true
                         });
            //  query = query.Where(x => !faze.Contains(x.StatusPhase));
            query = FilterReservations(criteria, ref query, ref count);

            return query;
```

**Comment by sdanyliv** [comment]
I think I've got your problem. Try to replace `short.Parse(f.Oznaka)` with `Sql.ConvertTo<short>.From(f.Oznaka)`.

**Comment by vladop** [comment]
I quickly tried and I think this is not problem. Also I have some filtering logic (additional part of query) which I will post when I have time. In some other queries I had problem that I was not able to use DateTime.Now.Year inside query so I had to make variable todaysYear=DateTime.Now.Year and use that variable inside query. All of this was working in BL.Toolkit

**Comment by MaceWindu** [comment]
Was fixed in 3.2.0
